Add "Verifying your contracts" guide (#2960)
NomicFoundation/hardhat: docs/src/content/hardhat-runner/docs/guides/verifying.md

# Verifying your contracts

Once your contract is ready, the next step is to deploy it to a live network and verify its source code.

> Verifying a contract means making public its source code and the compiler settings you used, which allows anyone to compile it and compare the generated bytecode with the one that is deployed on-chain. Doing this is extremely important in an open platform like Ethereum.

Reviewer: ```suggestion
Verifying a contract means making its source code public, along with the compiler settings you used, which allows anyone to compile it and compare the generated bytecode with the one that is deployed on-chain. Doing this is extremely important in an open platform like Ethereum.
```

> In this guide we'll explain how to do this in the [Etherscan](https://etherscan.io/) explorer, but there are other ways to verify a contract like [Sourcify](https://sourcify.dev/).

Reviewer: ```suggestion
In this guide we'll explain how to do this in the [Etherscan](https://etherscan.io/) explorer, but there are other ways to verify a contract, for example with [Sourcify](https://sourcify.dev/).
```

## Getting an API key from Etherscan

> The first thing you need to is an API key from Etherscan. To get one, go to [their site](https://etherscan.io/login), sign in (or create an account if you don't have one) and open the "API Keys" tab. Then click the "Add" button and give a name (like "Hardhat") to the API key you are creating. After that you'll see the newly created key in the list.

Reviewer: ```suggestion
The first thing you need is an API key from Etherscan. To get one, go to [their site](https://etherscan.io/login), sign in (or create an account if you don't have one) and open the "API Keys" tab. Then click the "Add" button and give a name (like "Hardhat") to the API key you are creating. After that you'll see the newly created key in the list.
```

Open your Hardhat config and add the API key you just created:

::::tabsgroup{options=TypeScript,JavaScript}

:::tab{value=TypeScript}

```ts
export default {
  // ...rest of the config...
  etherscan: {
    apiKey: "ABCDE12345ABCDE12345ABCDE123456789",
  },
};
```

:::

:::tab{value=JavaScript}

```js
module.exports = {
  // ...rest of the config...
  etherscan: {
    apiKey: "ABCDE12345ABCDE12345ABCDE123456789",
  },
};
```

:::

::::

## Deploying and verifying a contract in the Goerli testnet

We are going to use the [Goerli testnet](https://ethereum.org/en/developers/docs/networks/#goerli) to deploy and verify our contract, so you need to add this network in your Hardhat config. Here we are using [Alchemy](https://www.alchemy.com/) to connect to the network, but you can use any JSON-RPC URL you want.

::::tabsgroup{options=TypeScript,JavaScript}

:::tab{value=TypeScript}

```ts
// Go to https://www.alchemyapi.io, sign up, create
// a new App in its dashboard, and replace "KEY" with its key
const ALCHEMY_API_KEY = "KEY";

> // Replace this private key with your Goerli account private key

Reviewer: ```suggestion
// Replace this private key with your Goerli account private key.
```

// To export your private key from Metamask, open Metamask and

> // go to Account Details > Export Private Key

Reviewer: ```suggestion
// go to Account Details > Export Private Key.
```
